### PR TITLE
Implement cache flush for most adapters

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes
#11565 added support for flushing the full cache and is only implemented for Memcache(d) right now.  Implement it for more of the adapters.

### Testing Instructions
Calling the `flush()` method of the cache storage adapter causes the cache store to be flushed.  `JFactory::getCache()->flush()`.

### Documentation Changes Required
N/A

### CODE REVIEW

File: libraries/joomla/cache/storage/file.php [hunk 1]
```diff
@@ -209,6 +209,25 @@ public function clean($group, $mode = null)
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function flush()
+	{
+		$return = true;
+
+		foreach ($this->_folders($this->_root) as $folder)
+		{
+			$return = $this->_deleteFolder($this->_root . '/' . $folder);
```
Comment (sovainfo) [inline libraries/joomla/cache/storage/file.php]: Suspect this to need $return |= $this.....

**Comment by csthomas** [comment]
We could use `flush` method to `Delete All`.
backend_fails_304

**Comment by andrepereiradasilva** [comment]
shouldn't we have unit tests for this? or for another PR?

**Comment by alikon** [comment]
i've tested with  redis  using `JFactory::getCache()->_getStorage()->flush();`

since `FLUSHDB` 

> Delete all the keys of the currently selected DB

should be used very carefully

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on a1b785bd850d76e3d1bae0e1f698699af871a8a1<br /><br />with redis<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11898">issues.joomla.org/joomla-cms/11898</a>.</sub>

**Comment by andrepereiradasilva** [comment]
Hum so could this have  a side effect and flush sessions too?

**Comment by csthomas** [comment]
http://stackoverflow.com/questions/13968350/session-flush-is-not-working-in-magento-for-memcache

**Comment by mbabker** [comment]
> Hum so could this have a side effect and flush sessions too?

They all could and probably will have side effects honestly.  Unless we force our code to use different servers for cache and session or write the methods to loop through the entire cache store and only remove stuff explicitly added by the cache handler, it's going to delete everything as instructed by each handler's flush API (the only one without side effects is the filesystem handler unless you've got your sessions configured to use the cache directory, in which case, I'm only going to sit back and laugh when you come up with a bug report about that scenario).

**Comment by mbabker** [comment]
> shouldn't we have unit tests for this?

Done.

**Comment by andrepereiradasilva** [comment]
> (the only one without side effects is the filesystem handler unless you've got your sessions configured to use the cache directory, in which case, I'm only going to sit back and laugh when you come up with a bug report about that scenario).

ehehehhe

**Comment by csthomas** [comment]
I'm wondering whether `flush()` method should be more implemented.

May be we can deprecate `clean()` method and introduce `clear()` which can delete also all related vars(indexes, locks) from joomla, means safe `flush()`.

clear() = redesign 'clean()' + safe 'flush()'.

**Comment by csthomas** [comment]
I have tested this item :white_check_mark: successfully on 25d6335ed4ba8c59bb041065fec23f8e45f15685<br /><br />I have prepared console.php file to include joomla dependencies.

```
$ php -a -dauto_prepend_file=console.php
Interactive mode enabled

php > $cache = JFactory::getCache('', 'output');
php > print_r($cache->options);
Array
(
    [cachebase] => [...]/demo/cache
    [lifetime] => 5
    [language] => en-GB
    [storage] => file
    [defaultgroup] => 
    [locking] => 1
    [locktime] => 15
    [checkTime] => 1
    [caching] => 1
)
php > var_dump($cache->_getStorage()->getAll());
array(0) {
}
php > var_dump($cache->store('some data', 'id', 'group'));
bool(true)
php > var_dump($cache->store('some data2', 'id2', 'group2'));
bool(true)
php > var_dump($cache->_getStorage()->getAll());
array(2) {
  ["group2"]=>
  object(JCacheStorageHelper)#23 (3) {
    ["group"]=>
    string(6) "group2"
    ["size"]=>
    string(4) "0.08"
    ["count"]=>
    int(2)
  }
  ["group"]=>
  object(JCacheStorageHelper)#24 (3) {
    ["group"]=>
    string(5) "group"
    ["size"]=>
    string(4) "0.08"
    ["count"]=>
    int(2)
  }
}
php > var_dump($cache->_getStorage()->flush());
bool(true)
php > var_dump($cache->_getStorage()->getAll());
array(0) {
}

## ADDITIONAL TEST RELATED WITH https://github.com/joomla/joomla-cms/pull/10767
 
php > var_dump($cache->_getStorage()->get('id', 'group'));
bool(false)
php > var_dump($cache->_getStorage()->getAll());
array(1) {
  ["group"]=>
  object(JCacheStorageHelper)#21 (3) {
    ["group"]=>
    string(5) "group"
    ["size"]=>
    string(4) "0.03"
    ["count"]=>
    int(1)
  }
}
php > var_dump($cache->_getStorage()->flush());
bool(true)
php > var_dump($cache->_getStorage()->getAll());
array(0) {
}

## TEST APCU

php > $cache->options['storage'] = 'apcu';
php > var_dump($cache->_getStorage()->getAll());
array(0) {
}
php > var_dump($cache->store('some data', 'id', 'group'));   
bool(true)
php > var_dump($cache->store('some data2', 'id2', 'group2'));
bool(true)
php > var_dump($cache->_getStorage()->getAll());
array(2) {
  ["group2"]=>
  object(JCacheStorageHelper)#22 (3) {
    ["group"]=>
    string(6) "group2"
    ["size"]=>
    string(4) "0.64"
    ["count"]=>
    int(1)
  }
  ["group"]=>
  object(JCacheStorageHelper)#23 (3) {
    ["group"]=>
    string(5) "group"
    ["size"]=>
    string(4) "0.64"
    ["count"]=>
    int(1)
  }
}
php > var_dump($cache->_getStorage()->flush());
bool(true)
php > var_dump($cache->_getStorage()->getAll());
array(0) {
}
```<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11898">issues.joomla.org/joomla-cms/11898</a>.</sub>

**Comment by mbabker** [comment]
... well?

File: libraries/joomla/cache/cache.php [hunk 2]
```diff
@@ -316,6 +316,26 @@ public function clean($group = null, $mode = 'group')
...
+	 */
+	public function flush()
+	{
+		// Get the storage
+		$handler = $this->_getStorage();
+
+		if (!($handler instanceof Exception))
+		{
+			return $handler->flush();
+		}
```
Comment (csthomas) [inline libraries/joomla/cache/cache.php]: Can you rebase and remove above old "exception" stuff.

Comment (mbabker) [reply libraries/joomla/cache/cache.php]: Done

**Comment by csthomas** [comment]
I have tested this item :white_check_mark: successfully on 58f61f5b48ef11a01147d46391cfa1a0e2e9f25b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11898">issues.joomla.org/tracker/joomla-cms/11898</a>.</sub>

**Comment by brianteeman** [comment]
Any reason this cant be merged? Looks like we have two tests from @alikon and @csthomas

**Comment by mbabker** [comment]
Well, if someone wants to do one more test/review and then anyone but me hit merge then it's fine.

**Comment by mbabker** [comment]
No interest, fine.

**Comment by tonypartridge** [comment]
can you reopen I’ll test, sorry didn’t see this.

**Comment by ghost** [comment]
can't reopen, get "422 RuntimeException". Guess the PR is deleted.

**Pull request closed without merging**